src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -843,6 +947,33 @@ private static ComWrappers.ComInterfaceEntry[] LookupVtableEntries(Type type)
                 source.AppendLine($$"""
                                                 return default;
                                             }
+                                    """);
+
+                if (hasRuntimeClasNameEntries)
+                {
+                    source.AppendLine($$"""
+                                private static string LookupRuntimeClassName(Type type)
+                                {
+                                    string typeName = type.ToString();
+                                """);
+
+                    foreach (var vtableAttribute in value.vtableAttributes.ToImmutableHashSet().Where(v => !string.IsNullOrEmpty(v.RuntimeClassName)))
+                    {
+                        source.AppendLine($$"""
+                                if (typeName == "{{vtableAttribute.VtableLookupClassName}}")

```

Comment:
Just so I understand, are we switching on the type name here and not just on the type itself to avoid rooting the types? Otherwise we do we round-trip to `string` rather than comparing `Type` instances directly?

Reply:
Rooting might be one scenario, but if I recall the main scenario I did this before was to avoid triggering the static constructors on those types unnecessarily during the lookup.

Reply:
> I did this before was to avoid triggering the static constructors on those types unnecessarily during the lookup.

It is unexpected that `typeof(X)` would trigger a static constructor by itself. Where was this observed?

Reply:
You are right, it wasn't the static constructor.  Looking back at the history, it was the module initializer when the types came from different modules.  We were registering the lookup table as part of the module initializer and when we were going through the lookup table, we triggered other module initializers that triggered registration of those lookup tables and thereby the lock we had.

https://github.com/microsoft/CsWinRT/pull/1377#issuecomment-1808497979

https://github.com/microsoft/CsWinRT/pull/1377/commits/c9b190c1381d8f1bf9159a589b4ed0cde7fc9028

FWIW, the other scenario this allows us to handle is nested private classes (i.e. `typeof(global::System.Collections.Generic.List<global::TestClass.NestedClass>)`) as that is the first issue I hit when I tried an experiment to go back to not using strings to confirm the findings from before.

Reply:
Ooh that makes sense, the private nested type is actually one I can see would be pretty common. I also have a few of those types in my ComputeSharp WinRT component, for some types implementing interop COM types from Win2D interfaces needed for custom effects. Perhaps at least leave a detailed comment over that code (in the generator code, not in the actual generated code) to explain all this for future reference?